// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/webhook/rotate-webhook-secret.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = exports.Response = void 0;
const response_base_1 = require("../response-base");
/**
 * Parent response type for a RotateWebhookSecret request.  The
 * response object is resolved to a type-safe object of one of
 * the following subtypes:
 *
 * - {Success}
 * - {Error}
 *
 * `instanceof` type guards can be used to operate on the appropriate subtype.
 * @example
 * For example:
 * ```
 * if (response instanceof RotateWebhookSecret.Error) {
 *   // Handle error as appropriate.  The compiler will smart-cast `response` to type
 *   // `RotateWebhookSecret.Error` in this block, so you will have access to the properties
 *   // of the Error class; e.g. `response.errorCode()`.
 * }
 * ```
 */
class Response extends response_base_1.ResponseBase {
}
exports.Response = Response;
class _Success extends Response {
    constructor(props) {
        super();
        this._secret = props.secret;
        this._cacheName = props.cacheName;
        this._webhookName = props.webhookName;
    }
    secret() {
        return this._secret;
    }
    webhookName() {
        return this._webhookName;
    }
    cacheName() {
        return this._cacheName;
    }
}
/**
 * Indicates a Successful RotateWebhookSecret request.
 */
class Success extends (0, response_base_1.ResponseSuccess)(_Success) {
}
exports.Success = Success;
class _Error extends Response {
    constructor(_innerException) {
        super();
        this._innerException = _innerException;
    }
}
/**
 * Indicates that an error occurred during the RotateWebhookSecret request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends (0, response_base_1.ResponseError)(_Error) {
}
exports.Error = Error;
