// llamaindex@0.2.1/dist/cjs/llm/base.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "BaseLLM", {
    enumerable: true,
    get: function() {
        return BaseLLM;
    }
});
const _utils = require("./utils.js");
class BaseLLM {
    async complete(params) {
        const { prompt, parentEvent, stream } = params;
        if (stream) {
            const stream = await this.chat({
                messages: [
                    {
                        content: prompt,
                        role: "user"
                    }
                ],
                parentEvent,
                stream: true
            });
            return (0, _utils.streamConverter)(stream, (chunk)=>{
                return {
                    text: chunk.delta
                };
            });
        }
        const chatResponse = await this.chat({
            messages: [
                {
                    content: prompt,
                    role: "user"
                }
            ],
            parentEvent
        });
        return {
            text: chatResponse.message.content
        };
    }
}
