// @gomomento/sdk-core@1.68.1/dist/src/config/logging/noop-momento-logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NoopMomentoLoggerFactory = exports.NoopMomentoLogger = void 0;
class NoopMomentoLogger {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    debug(msg, ...args) {
        // no-op
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    error(msg, ...args) {
        // no-op
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    info(msg, ...args) {
        // no-op
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    trace(msg, ...args) {
        // no-op
    }
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    warn(msg, ...args) {
        // no-op
    }
}
exports.NoopMomentoLogger = NoopMomentoLogger;
class NoopMomentoLoggerFactory {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars, @typescript-eslint/no-explicit-any
    getLogger(loggerName) {
        return new NoopMomentoLogger();
    }
}
exports.NoopMomentoLoggerFactory = NoopMomentoLoggerFactory;
