// @gomomento/sdk@1.68.1/dist/src/config/leaderboard-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LeaderboardClientConfiguration = void 0;
class LeaderboardClientConfiguration {
    constructor(props) {
        this.loggerFactory = props.loggerFactory;
        this.transportStrategy = props.transportStrategy;
        this.throwOnErrors = props.throwOnErrors;
        this.middlewares = props.middlewares;
    }
    getLoggerFactory() {
        return this.loggerFactory;
    }
    getTransportStrategy() {
        return this.transportStrategy;
    }
    withClientTimeoutMillis(clientTimeoutMillis) {
        return new LeaderboardClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: this.transportStrategy.withClientTimeoutMillis(clientTimeoutMillis),
            throwOnErrors: this.throwOnErrors,
            middlewares: this.middlewares,
        });
    }
    withTransportStrategy(transportStrategy) {
        return new LeaderboardClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: transportStrategy,
            throwOnErrors: this.throwOnErrors,
            middlewares: this.middlewares,
        });
    }
    getThrowOnErrors() {
        return this.throwOnErrors;
    }
    withThrowOnErrors(throwOnErrors) {
        return new LeaderboardClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: this.transportStrategy,
            throwOnErrors: throwOnErrors,
            middlewares: this.middlewares,
        });
    }
    getMiddlewares() {
        return this.middlewares;
    }
    withMiddlewares(middlewares) {
        return new LeaderboardClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: this.transportStrategy,
            middlewares: middlewares,
            throwOnErrors: this.throwOnErrors,
        });
    }
    addMiddleware(middleware) {
        return new LeaderboardClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: this.transportStrategy,
            middlewares: [middleware, ...this.middlewares],
            throwOnErrors: this.throwOnErrors,
        });
    }
}
exports.LeaderboardClientConfiguration = LeaderboardClientConfiguration;
