// @gomomento/sdk@1.68.1/dist/src/config/vector-index-configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VectorIndexClientConfiguration = void 0;
class VectorIndexClientConfiguration {
    constructor(props) {
        this.loggerFactory = props.loggerFactory;
        this.transportStrategy = props.transportStrategy;
        this.throwOnErrors = props.throwOnErrors;
    }
    getLoggerFactory() {
        return this.loggerFactory;
    }
    getTransportStrategy() {
        return this.transportStrategy;
    }
    withClientTimeoutMillis(clientTimeoutMillis) {
        return new VectorIndexClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: this.transportStrategy.withClientTimeoutMillis(clientTimeoutMillis),
            throwOnErrors: this.throwOnErrors,
        });
    }
    withTransportStrategy(transportStrategy) {
        return new VectorIndexClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: transportStrategy,
            throwOnErrors: this.throwOnErrors,
        });
    }
    getThrowOnErrors() {
        return this.throwOnErrors;
    }
    withThrowOnErrors(throwOnErrors) {
        return new VectorIndexClientConfiguration({
            loggerFactory: this.loggerFactory,
            transportStrategy: this.transportStrategy,
            throwOnErrors: throwOnErrors,
        });
    }
}
exports.VectorIndexClientConfiguration = VectorIndexClientConfiguration;
