// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/cache-list-fetch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Miss = exports.Hit = exports.Response = void 0;
const response_base_1 = require("./response-base");
const utils_1 = require("../../internal/utils");
const TEXT_DECODER = new TextDecoder();
/**
 * Parent response type for a list fetch request.  The
 * response object is resolved to a type-safe object of one of
 * the following subtypes:
 *
 * - {Hit}
 * - {Miss}
 * - {Error}
 *
 * `instanceof` type guards can be used to operate on the appropriate subtype.
 * @example
 * For example:
 * ```
 * if (response instanceof CacheListFetch.Error) {
 *   // Handle error as appropriate.  The compiler will smart-cast `response` to type
 *   // `CacheListFetch.Error` in this block, so you will have access to the properties
 *   // of the Error class; e.g. `response.errorCode()`.
 * }
 * ```
 */
class Response extends response_base_1.ResponseBase {
    value() {
        if (this instanceof Hit) {
            return this.value();
        }
        return undefined;
    }
}
exports.Response = Response;
class _Hit extends Response {
    constructor(values) {
        super();
        this._values = values;
    }
    /**
     * Returns the data as an array of byte arrays.
     * @returns {Uint8Array[]}
     */
    valueListUint8Array() {
        return this._values;
    }
    /**
     * Returns the data as an array of strings, decoded from the underlying byte array.
     * @returns {string[]}
     */
    valueListString() {
        return this._values.map(v => TEXT_DECODER.decode(v));
    }
    /**
     * Returns the data as an array of strings, decoded from the underlying byte array.  This is a convenience alias
     * for {valueListString}
     * @returns {string[]}
     */
    valueList() {
        return this.valueListString();
    }
    /**
     * Returns the data as an array of strings, decoded from the underlying byte array.  This is a convenience alias
     * for {valueListString}
     * @returns {string[]}
     */
    value() {
        return this.valueListString();
    }
    toString() {
        const truncatedStringArray = (0, utils_1.truncateStringArray)(this.valueListString());
        return `${super.toString()}: [${truncatedStringArray.toString()}]`;
    }
}
/**
 * Indicates that the requested data was successfully retrieved from the cache.  Provides
 * `value*` accessors to retrieve the data in the appropriate format.
 */
class Hit extends (0, response_base_1.ResponseHit)(_Hit) {
}
exports.Hit = Hit;
class _Miss extends Response {
}
/**
 * Indicates that the requested data was not available in the cache.
 */
class Miss extends (0, response_base_1.ResponseMiss)(_Miss) {
}
exports.Miss = Miss;
class _Error extends Response {
    constructor(_innerException) {
        super();
        this._innerException = _innerException;
    }
}
/**
 * Indicates that an error occurred during the cache delete request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends (0, response_base_1.ResponseError)(_Error) {
}
exports.Error = Error;
