// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/cache-batch-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = exports.Response = void 0;
const response_base_1 = require("./response-base");
const TEXT_DECODER = new TextDecoder();
/**
 * Parent response type for a cache get batch request.  The
 * response object is resolved to a type-safe object of one of
 * the following subtypes:
 *
 * - {Success}
 * - {Error}
 *
 * `instanceof` type guards can be used to operate on the appropriate subtype.
 * @example
 * For example:
 * ```
 * if (response instanceof BatchGet.Error) {
 *   // Handle error as appropriate.  The compiler will smart-cast `response` to type
 *   // `BatchGet.Error` in this block, so you will have access to the properties
 *   // of the Error class; e.g. `response.errorCode()`.
 * }
 * ```
 */
class Response extends response_base_1.ResponseBase {
    values() {
        if (this instanceof Success) {
            return this.values();
        }
        return undefined;
    }
    results() {
        if (this instanceof Success) {
            return this.results();
        }
        return undefined;
    }
}
exports.Response = Response;
class _Success extends Response {
    constructor(items, keys) {
        super();
        this.items = items;
        this.keys = keys;
    }
    /**
     * Returns the status for each request in the batch as a list of CacheGet.Response objects.
     * @returns {CacheGet.Response[]}
     */
    results() {
        return this.items;
    }
    /**
     * Returns the data as a Record whose keys and values are utf-8 strings, decoded from the underlying byte arrays.
     * This can be used in most places where an Object is desired.  This is a convenience alias for
     * {valueRecordStringString}.
     * @returns {Record<string, string>}
     */
    values() {
        return this.valuesRecordStringString();
    }
    /**
     * Returns the data as a Record whose keys and values are utf-8 strings, decoded from the underlying byte arrays.
     * This can be used in most places where an Object is desired.  This is a convenience alias for
     * {valueRecordStringString}.
     * @returns {Record<string, string>}
     */
    valuesRecord() {
        return this.valuesRecordStringString();
    }
    /**
     * Returns the data as a Record whose keys and values are utf-8 strings, decoded from the underlying byte arrays.
     * This can be used in most places where an Object is desired.
     * @returns {Record<string, string>}
     */
    valuesRecordStringString() {
        return this.items.reduce((acc, item, index) => {
            if (item.value() !== undefined) {
                acc[TEXT_DECODER.decode(this.keys[index])] = item.valueString();
            }
            return acc;
        }, {});
    }
    /**
     * Returns the data as a Record whose keys are utf-8 strings, decoded from the underlying byte array, and whose
     * values are byte arrays.  This can be used in most places where an Object is desired.
     * @returns {Record<string, Uint8Array>}
     */
    valuesRecordStringUint8Array() {
        return this.items.reduce((acc, item, index) => {
            if (item.value() !== undefined) {
                acc[TEXT_DECODER.decode(this.keys[index])] = item.valueUint8Array();
            }
            return acc;
        }, {});
    }
    /**
     * Returns the data as a Map whose keys and values are utf-8 strings, decoded from the underlying byte arrays.
     * This is a convenience alias for {valueMapStringString}.
     * @returns {Map<string, string>}
     */
    valuesMap() {
        return this.valuesMapStringString();
    }
    /**
     * Returns the data as a Map whose keys and values are utf-8 strings, decoded from the underlying byte arrays.
     * @returns {Map<string, string>}
     */
    valuesMapStringString() {
        return this.items.reduce((acc, item, index) => {
            if (item.value() !== undefined) {
                acc.set(TEXT_DECODER.decode(this.keys[index]), item.valueString());
            }
            return acc;
        }, new Map());
    }
    /**
     * Returns the data as a Map whose keys and values are byte arrays.
     * @returns {Map<Uint8Array, Uint8Array>}
     */
    valuesMapUint8ArrayUint8Array() {
        return this.items.reduce((acc, item, index) => {
            if (item.value() !== undefined) {
                acc.set(this.keys[index], item.valueUint8Array());
            }
            return acc;
        }, new Map());
    }
    toString() {
        const display = this.results()
            .map(result => result.toString())
            .toString();
        return `${super.toString()}: ${display}`;
    }
}
/**
 * Indicates that the requested data was successfully retrieved from the cache.  Provides
 * `value*` accessors to retrieve the data in the appropriate format.
 */
class Success extends (0, response_base_1.ResponseSuccess)(_Success) {
}
exports.Success = Success;
class _Error extends Response {
    constructor(_innerException) {
        super();
        this._innerException = _innerException;
    }
}
/**
 * Indicates that an error occurred during the cache get request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends (0, response_base_1.ResponseError)(_Error) {
}
exports.Error = Error;
