// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/create-signing-key.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = exports.Response = void 0;
const response_base_1 = require("./response-base");
/**
 * Parent response type for a create signing key request.  The
 * response object is resolved to a type-safe object of one of
 * the following subtypes:
 *
 * - {Success}
 * - {Error}
 *
 * `instanceof` type guards can be used to operate on the appropriate subtype.
 * @example
 * For example:
 * ```
 * if (response instanceof CreateSigningKey.Error) {
 *   // Handle error as appropriate.  The compiler will smart-cast `response` to type
 *   // `CreateSigningKey.Error` in this block, so you will have access to the properties
 *   // of the Error class; e.g. `response.errorCode()`.
 * }
 * ```
 */
class Response extends response_base_1.ResponseBase {
}
exports.Response = Response;
class _Success extends Response {
    constructor(endpoint, result) {
        var _a, _b;
        super();
        const key = (_a = result === null || result === void 0 ? void 0 : result.key) !== null && _a !== void 0 ? _a : '';
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
        this.keyId = JSON.parse(key)['kid'];
        this.endpoint = endpoint;
        this.key = key;
        this.expiresAt = new Date((_b = result === null || result === void 0 ? void 0 : result.expiresAt) !== null && _b !== void 0 ? _b : 0);
    }
    getKeyId() {
        return this.keyId;
    }
    getEndpoint() {
        return this.endpoint;
    }
    getKey() {
        return this.key;
    }
    getExpiresAt() {
        return this.expiresAt;
    }
}
/**
 * Indicates a Successful create signing key request.
 */
class Success extends (0, response_base_1.ResponseSuccess)(_Success) {
}
exports.Success = Success;
class _Error extends Response {
    constructor(_innerException) {
        super();
        this._innerException = _innerException;
    }
}
/**
 * Indicates that an error occurred during the create signing key request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends (0, response_base_1.ResponseError)(_Error) {
}
exports.Error = Error;
