// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/grpc-response-types.js
"use strict";
/**
 * This file exists as a layer between our javascript classes and the grpc generated
 * ones. Its required since the types generated by grpc-web are different than those
 * generated by grpc for node. Inside of the respective clients, we will map the grpc
 * response types to these corresponding types. This allows us to share the response
 * messages between the 2 sdks.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports._RankedElement = exports._SortedSetElement = exports._SigningKey = exports._SortedSetGetScoreResponsePart = exports._DictionaryGetResponsePart = exports._DictionaryFieldValuePair = exports._ECacheResult = void 0;
var _ECacheResult;
(function (_ECacheResult) {
    _ECacheResult[_ECacheResult["Invalid"] = 0] = "Invalid";
    _ECacheResult[_ECacheResult["Ok"] = 1] = "Ok";
    _ECacheResult[_ECacheResult["Hit"] = 2] = "Hit";
    _ECacheResult[_ECacheResult["Miss"] = 3] = "Miss";
})(_ECacheResult = exports._ECacheResult || (exports._ECacheResult = {}));
class _DictionaryFieldValuePair {
    constructor({ field, value }) {
        this.field = field;
        this.value = value;
    }
}
exports._DictionaryFieldValuePair = _DictionaryFieldValuePair;
class _DictionaryGetResponsePart {
    constructor(result, cacheBody) {
        this.result = result;
        this.cacheBody = cacheBody;
    }
}
exports._DictionaryGetResponsePart = _DictionaryGetResponsePart;
class _SortedSetGetScoreResponsePart {
    constructor(result, score) {
        this.result = result;
        this.score = score;
    }
}
exports._SortedSetGetScoreResponsePart = _SortedSetGetScoreResponsePart;
class _SigningKey {
    constructor(key, expiresAt) {
        this.key = key !== null && key !== void 0 ? key : '';
        this.expiresAt = expiresAt !== null && expiresAt !== void 0 ? expiresAt : 0;
    }
}
exports._SigningKey = _SigningKey;
class _SortedSetElement {
    constructor(value, score) {
        this.value = value;
        this.score = score;
    }
}
exports._SortedSetElement = _SortedSetElement;
class _RankedElement {
    constructor(id, score, rank) {
        this.id = id;
        this.score = score;
        this.rank = rank;
    }
}
exports._RankedElement = _RankedElement;
