// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/list-signing-keys.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = exports.Response = void 0;
const signing_key_1 = require("../signing-key");
const response_base_1 = require("./response-base");
/**
 * Parent response type for a list signing keys request.  The
 * response object is resolved to a type-safe object of one of
 * the following subtypes:
 *
 * - {Success}
 * - {Error}
 *
 * `instanceof` type guards can be used to operate on the appropriate subtype.
 * @example
 * For example:
 * ```
 * if (response instanceof ListSigningKeys.Error) {
 *   // Handle error as appropriate.  The compiler will smart-cast `response` to type
 *   // `ListSigningKeys.Error` in this block, so you will have access to the properties
 *   // of the Error class; e.g. `response.errorCode()`.
 * }
 * ```
 */
class Response extends response_base_1.ResponseBase {
}
exports.Response = Response;
class _Success extends Response {
    constructor(endpoint, signingKeys, nextToken) {
        var _a;
        super();
        this.nextToken = nextToken;
        this.signingKeys =
            (_a = signingKeys.map(signingKey => new signing_key_1.SigningKey(signingKey.key, new Date(signingKey.expiresAt * 1000), endpoint))) !== null && _a !== void 0 ? _a : [];
    }
    getNextToken() {
        return this.nextToken;
    }
    getSigningKeys() {
        return this.signingKeys;
    }
}
/**
 * Indicates a Successful list signing keys request.
 */
class Success extends (0, response_base_1.ResponseSuccess)(_Success) {
}
exports.Success = Success;
class _Error extends Response {
    constructor(_innerException) {
        super();
        this._innerException = _innerException;
    }
}
/**
 * Indicates that an error occurred during the list signing keys request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends (0, response_base_1.ResponseError)(_Error) {
}
exports.Error = Error;
