// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/topic-subscribe.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Subscription = exports.Response = void 0;
const response_base_1 = require("./response-base");
/**
 * Parent response type for a cache get request.  The
 * response object is resolved to a type-safe object of one of
 * the following subtypes:
 *
 * - {Subscription}
 * - {Item}
 * - {Error}
 *
 * `instanceof` type guards can be used to operate on the appropriate subtype.
 * @example
 * For example:
 * ```
 * if (response instanceof TopicSubscribe.Error) {
 *   // Handle error as appropriate.  The compiler will smart-cast `response` to type
 *   // `CacheGet.Error` in this block, so you will have access to the properties
 *   // of the Error class; e.g. `response.errorCode()`.
 * }
 * ```
 */
class Response extends response_base_1.ResponseBase {
}
exports.Response = Response;
/**
 * Encapsulates a topic subscription.
 *
 * @remarks Currently allows unsubscribing from the topic.
 * In the future, this may be extended to include additional
 * statistics about the subscription.
 */
class Subscription extends Response {
    constructor(subscriptionState) {
        super();
        this.subscriptionState = subscriptionState;
    }
    /**
     * Unsubscribes from the topic.
     *
     * @returns void
     */
    unsubscribe() {
        this.subscriptionState.unsubscribe();
    }
    get isSubscribed() {
        return this.subscriptionState.isSubscribed;
    }
}
exports.Subscription = Subscription;
class _Error extends Response {
    constructor(_innerException) {
        super();
        this._innerException = _innerException;
    }
}
/**
 * Indicates that an error occurred during the topic subscribe request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends (0, response_base_1.ResponseError)(_Error) {
}
exports.Error = Error;
