// @gomomento/sdk-core@1.68.1/dist/src/auth/tokens/disposable-token-scopes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.cacheKeyPrefixWriteOnly = exports.cacheKeyWriteOnly = exports.cacheKeyPrefixReadOnly = exports.cacheKeyReadOnly = exports.cacheKeyPrefixReadWrite = exports.cacheKeyReadWrite = void 0;
const permission_scope_1 = require("./permission-scope");
__exportStar(require("./permission-scopes"), exports);
function cacheKeyReadWrite(cacheSelector, key) {
    return {
        permissions: [{ role: permission_scope_1.CacheRole.ReadWrite, cache: cacheSelector, item: key }],
    };
}
exports.cacheKeyReadWrite = cacheKeyReadWrite;
function cacheKeyPrefixReadWrite(cacheSelector, keyPrefix) {
    return {
        permissions: [
            {
                role: permission_scope_1.CacheRole.ReadWrite,
                cache: cacheSelector,
                item: { keyPrefix: keyPrefix },
            },
        ],
    };
}
exports.cacheKeyPrefixReadWrite = cacheKeyPrefixReadWrite;
function cacheKeyReadOnly(cacheSelector, key) {
    return {
        permissions: [{ role: permission_scope_1.CacheRole.ReadOnly, cache: cacheSelector, item: key }],
    };
}
exports.cacheKeyReadOnly = cacheKeyReadOnly;
function cacheKeyPrefixReadOnly(cacheSelector, keyPrefix) {
    return {
        permissions: [
            {
                role: permission_scope_1.CacheRole.ReadOnly,
                cache: cacheSelector,
                item: { keyPrefix: keyPrefix },
            },
        ],
    };
}
exports.cacheKeyPrefixReadOnly = cacheKeyPrefixReadOnly;
function cacheKeyWriteOnly(cacheSelector, key) {
    return {
        permissions: [{ role: permission_scope_1.CacheRole.WriteOnly, cache: cacheSelector, item: key }],
    };
}
exports.cacheKeyWriteOnly = cacheKeyWriteOnly;
function cacheKeyPrefixWriteOnly(cacheSelector, keyPrefix) {
    return {
        permissions: [
            {
                role: permission_scope_1.CacheRole.WriteOnly,
                cache: cacheSelector,
                item: { keyPrefix: keyPrefix },
            },
        ],
    };
}
exports.cacheKeyPrefixWriteOnly = cacheKeyPrefixWriteOnly;
