// @gomomento/sdk-core@1.68.1/dist/src/auth/tokens/disposable-token-scope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.asDisposableTokenPermissionsObject = exports.isDisposableTokenPermissionsObject = exports.asDisposableTokenCachePermission = exports.isDisposableTokenCachePermission = exports.isCacheItemKeyPrefix = exports.isCacheItemKey = void 0;
const permission_scope_1 = require("./permission-scope");
function isCacheItemKey(cacheItem) {
    if (cacheItem === permission_scope_1.AllCacheItems) {
        return false;
    }
    if (typeof cacheItem === 'string') {
        return true;
    }
    return 'key' in cacheItem;
}
exports.isCacheItemKey = isCacheItemKey;
function isCacheItemKeyPrefix(cacheItem) {
    if (cacheItem === permission_scope_1.AllCacheItems) {
        return false;
    }
    if (typeof cacheItem === 'string') {
        return false;
    }
    return 'keyPrefix' in cacheItem;
}
exports.isCacheItemKeyPrefix = isCacheItemKeyPrefix;
function isDisposableTokenCachePermission(p) {
    return 'role' in p && 'cache' in p && 'item' in p && !('topic' in p);
}
exports.isDisposableTokenCachePermission = isDisposableTokenCachePermission;
function asDisposableTokenCachePermission(p) {
    if (!isDisposableTokenCachePermission(p)) {
        throw new Error(`permission is not a DisposableTokenCachePermission object: ${JSON.stringify(p)}`);
    }
    return p;
}
exports.asDisposableTokenCachePermission = asDisposableTokenCachePermission;
function isDisposableTokenPermissionObject(p) {
    return isDisposableTokenCachePermission(p);
}
function isDisposableTokenPermissionsObject(scope) {
    if (!('permissions' in scope)) {
        return false;
    }
    const permissions = scope.permissions;
    return permissions.every(p => isDisposableTokenPermissionObject(p));
}
exports.isDisposableTokenPermissionsObject = isDisposableTokenPermissionsObject;
function asDisposableTokenPermissionsObject(scope) {
    if (!isDisposableTokenPermissionsObject(scope)) {
        throw new Error(`Token scope is not a DisposableTokenCachePermissions object: ${JSON.stringify(scope)}`);
    }
    return scope;
}
exports.asDisposableTokenPermissionsObject = asDisposableTokenPermissionsObject;
