// @gomomento/sdk-core@1.68.1/dist/src/auth/tokens/permission-scopes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.topicPublishOnly = exports.topicPublishSubscribe = exports.topicSubscribeOnly = exports.cacheWriteOnly = exports.cacheReadOnly = exports.cacheReadWrite = void 0;
const permission_scope_1 = require("./permission-scope");
function cacheReadWrite(cacheSelector) {
    return {
        permissions: [{ role: permission_scope_1.CacheRole.ReadWrite, cache: cacheSelector }],
    };
}
exports.cacheReadWrite = cacheReadWrite;
function cacheReadOnly(cacheSelector) {
    return {
        permissions: [{ role: permission_scope_1.CacheRole.ReadOnly, cache: cacheSelector }],
    };
}
exports.cacheReadOnly = cacheReadOnly;
function cacheWriteOnly(cacheSelector) {
    return {
        permissions: [{ role: permission_scope_1.CacheRole.WriteOnly, cache: cacheSelector }],
    };
}
exports.cacheWriteOnly = cacheWriteOnly;
function topicSubscribeOnly(cacheSelector, topicSelector) {
    return {
        permissions: [
            {
                role: permission_scope_1.TopicRole.SubscribeOnly,
                cache: cacheSelector,
                topic: topicSelector,
            },
        ],
    };
}
exports.topicSubscribeOnly = topicSubscribeOnly;
function topicPublishSubscribe(cacheSelector, topicSelector) {
    return {
        permissions: [
            {
                role: permission_scope_1.TopicRole.PublishSubscribe,
                cache: cacheSelector,
                topic: topicSelector,
            },
        ],
    };
}
exports.topicPublishSubscribe = topicPublishSubscribe;
function topicPublishOnly(cacheSelector, topicSelector) {
    return {
        permissions: [
            {
                role: permission_scope_1.TopicRole.PublishOnly,
                cache: cacheSelector,
                topic: topicSelector,
            },
        ],
    };
}
exports.topicPublishOnly = topicPublishOnly;
