// @gomomento/sdk-core@1.68.1/dist/src/auth/tokens/permission-scope.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PredefinedScope = exports.asPermissionsObject = exports.isPermissionsObject = exports.AllDataReadWrite = exports.asTopicPermission = exports.isTopicPermission = exports.isTopicName = exports.TopicRole = exports.asCachePermission = exports.isCachePermission = exports.isCacheName = exports.AllCacheItems = exports.AllTopics = exports.AllCaches = exports.CacheRole = void 0;
var CacheRole;
(function (CacheRole) {
    CacheRole["ReadWrite"] = "readwrite";
    CacheRole["ReadOnly"] = "readonly";
    CacheRole["WriteOnly"] = "writeonly";
})(CacheRole = exports.CacheRole || (exports.CacheRole = {}));
exports.AllCaches = Symbol();
exports.AllTopics = Symbol();
exports.AllCacheItems = Symbol();
function isCacheName(cache) {
    if (cache === exports.AllCaches) {
        return false;
    }
    return 'name' in cache;
}
exports.isCacheName = isCacheName;
function isCachePermission(p) {
    return 'role' in p && 'cache' in p && !('topic' in p);
}
exports.isCachePermission = isCachePermission;
function asCachePermission(p) {
    if (!isCachePermission(p)) {
        throw new Error(`permission is not a CachePermission object: ${JSON.stringify(p)}`);
    }
    return p;
}
exports.asCachePermission = asCachePermission;
var TopicRole;
(function (TopicRole) {
    TopicRole["PublishSubscribe"] = "publishsubscribe";
    TopicRole["SubscribeOnly"] = "subscribeonly";
    TopicRole["PublishOnly"] = "publishonly";
})(TopicRole = exports.TopicRole || (exports.TopicRole = {}));
function isTopicName(topic) {
    if (topic === exports.AllTopics) {
        return false;
    }
    return 'name' in topic;
}
exports.isTopicName = isTopicName;
function isTopicPermission(p) {
    return 'role' in p && 'cache' in p && 'topic' in p;
}
exports.isTopicPermission = isTopicPermission;
function asTopicPermission(p) {
    if (!isTopicPermission(p)) {
        throw new Error(`permission is not a TopicPermission object: ${JSON.stringify(p)}`);
    }
    return p;
}
exports.asTopicPermission = asTopicPermission;
exports.AllDataReadWrite = {
    permissions: [
        { role: CacheRole.ReadWrite, cache: exports.AllCaches },
        { role: TopicRole.PublishSubscribe, cache: exports.AllCaches, topic: exports.AllTopics },
    ],
};
function isPermissionObject(p) {
    return isCachePermission(p) || isTopicPermission(p);
}
function isPermissionsObject(scope) {
    if (!('permissions' in scope)) {
        return false;
    }
    const permissions = scope.permissions;
    return permissions.every(p => isPermissionObject(p));
}
exports.isPermissionsObject = isPermissionsObject;
function asPermissionsObject(scope) {
    if (!isPermissionsObject(scope)) {
        throw new Error(`Token scope is not a Permissions object: ${JSON.stringify(scope)}`);
    }
    return scope;
}
exports.asPermissionsObject = asPermissionsObject;
class PredefinedScope {
}
exports.PredefinedScope = PredefinedScope;
