// @gomomento/sdk-core@1.68.1/dist/src/messages/responses/vector/list-vector-indexes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Error = exports.Success = exports.Response = void 0;
const response_base_1 = require("../response-base");
/**
 * Parent response type for a list indexes request.  The
 * response object is resolved to a type-safe object of one of
 * the following subtypes:
 *
 * - {Success}
 * - {Error}
 *
 * `instanceof` type guards can be used to operate on the appropriate subtype.
 * @example
 * For example:
 * ```
 * if (response instanceof ListIndexes.Error) {
 *   // Handle error as appropriate.  The compiler will smart-cast `response` to type
 *   // `ListIndexes.Error` in this block, so you will have access to the properties
 *   // of the Error class; e.g. `response.errorCode()`.
 * }
 * ```
 */
class Response extends response_base_1.ResponseBase {
    getIndexes() {
        if (this instanceof Success) {
            return this.getIndexes();
        }
        return undefined;
    }
}
exports.Response = Response;
class _Success extends Response {
    constructor(indexes) {
        super();
        this.indexes = indexes;
    }
    /**
     * An array of index names.
     * @returns {string[]}
     */
    getIndexes() {
        return this.indexes;
    }
    toString() {
        const indexes = this.indexes.map(indexInfo => indexInfo.toString());
        return super.toString() + ': [' + indexes.join(', ') + ']';
    }
}
/**
 * Indicates a Successful list indexes request.
 */
class Success extends (0, response_base_1.ResponseSuccess)(_Success) {
}
exports.Success = Success;
class _Error extends Response {
    constructor(_innerException) {
        super();
        this._innerException = _innerException;
    }
}
/**
 * Indicates that an error occurred during the list indexes request.
 *
 * This response object includes the following fields that you can use to determine
 * how you would like to handle the error:
 *
 * - `errorCode()` - a unique Momento error code indicating the type of error that occurred.
 * - `message()` - a human-readable description of the error
 * - `innerException()` - the original error that caused the failure; can be re-thrown.
 */
class Error extends (0, response_base_1.ResponseError)(_Error) {
}
exports.Error = Error;
