// @gomomento/sdk@1.68.1/dist/src/internal/middleware/in-flight-request-count-middleware.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InFlightRequestCountMiddleware = void 0;
class InFlightRequestCountMiddlewareRequestHandler {
    constructor(logger) {
        this.logger = logger;
        this.inFlightRequestsAtStart =
            ++InFlightRequestCountMiddleware.numActiveRequests;
    }
    onRequestBody(request) {
        return Promise.resolve(request);
    }
    onRequestMetadata(metadata) {
        return Promise.resolve(metadata);
    }
    onResponseBody(response) {
        this.receivedResponseBody = true;
        if (this.done())
            this.logStatusInformation();
        return Promise.resolve(response);
    }
    onResponseMetadata(metadata) {
        return Promise.resolve(metadata);
    }
    onResponseStatus(status) {
        this.receivedResponseStatus = true;
        if (this.done())
            this.logStatusInformation();
        return Promise.resolve(status);
    }
    done() {
        return this.receivedResponseBody && this.receivedResponseStatus;
    }
    logStatusInformation() {
        const inFlightRequestsAtEnd = --InFlightRequestCountMiddleware.numActiveRequests;
        this.logger.debug(`Request completed; in-flight requests at start: ${this.inFlightRequestsAtStart}, in-flight requests at completion: ${inFlightRequestsAtEnd}`);
    }
}
class InFlightRequestCountMiddleware {
    constructor(loggerFactory) {
        this.logger = loggerFactory.getLogger(this);
    }
    onNewRequest() {
        return new InFlightRequestCountMiddlewareRequestHandler(this.logger);
    }
}
exports.InFlightRequestCountMiddleware = InFlightRequestCountMiddleware;
InFlightRequestCountMiddleware.numActiveRequests = 0;
