// @gomomento/sdk-core@1.68.1/dist/src/utils/expiration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExpiresAt = exports.ExpiresIn = void 0;
class Expiration {
    constructor(doesExpire) {
        this._doesExpire = doesExpire;
    }
    /**
     * Whether the token expires or not.
     * @returns {boolean}
     */
    doesExpire() {
        return this._doesExpire;
    }
}
class ExpiresIn extends Expiration {
    /**
     * If doesExpire is false, the refresh token will not have a expiration time, instead validForSeconds will be 'Infinity'.
     * @param {number} [validForSeconds]
     */
    constructor(validForSeconds) {
        super(validForSeconds !== Infinity);
        this._validForSeconds =
            validForSeconds === null ? Infinity : validForSeconds;
    }
    /**
     * Time token is valid for in seconds.
     * @returns {number} Infinity, if token doesn't expire.
     */
    seconds() {
        return this._validForSeconds;
    }
    /**
     * Constructs a ExpiresIn which never expires.
     * @returns {ExpiresIn}
     */
    static never() {
        return new ExpiresIn(Infinity);
    }
    /**
     * Constructs a ExpiresIn with a specified validFor period in seconds.
     * If seconds are undefined, or null, then token never expires.
     * @param validForSeconds
     * @returns {ExpiresIn}
     */
    static seconds(validForSeconds) {
        return new ExpiresIn(validForSeconds);
    }
    /**
     * Constructs a ExpiresIn with a specified validFor period in minutes.
     * @param validForMinutes
     * @returns {ExpiresIn}
     */
    static minutes(validForMinutes) {
        return new ExpiresIn(validForMinutes * 60);
    }
    /**
     * Constructs a ExpiresIn with a specified validFor period in hours.
     * @param validForHours
     * @returns {ExpiresIn}
     */
    static hours(validForHours) {
        return new ExpiresIn(validForHours * 3600);
    }
    /**
     * Constructs an ExpiresIn with a specified validFor period in days.
     * @param validForDays
     * @returns {ExpiresIn}
     */
    static days(validForDays) {
        return new ExpiresIn(validForDays * 86400);
    }
    /**
     * Constructs an ExpiresIn with a specified expiresBy period in epoch format.
     * @param expiresBy
     * @returns {ExpiresIn}
     */
    static epoch(expiresBy) {
        const currentEpoch = new Date().getTime() / 1000;
        const secondsUntilEpoch = Math.round(expiresBy - currentEpoch);
        return new ExpiresIn(secondsUntilEpoch);
    }
}
exports.ExpiresIn = ExpiresIn;
class ExpiresAt extends Expiration {
    constructor(epochTimestamp) {
        super(epochTimestamp !== undefined && epochTimestamp !== 0);
        if (this.doesExpire()) {
            this.validUntil = epochTimestamp;
        }
        else {
            this.validUntil = Infinity;
        }
    }
    /**
     * Returns epoch timestamp of when api token expires.
     * @returns {number}
     */
    epoch() {
        return this.validUntil;
    }
    /**
     * Constructs an ExpiresAt with the specified epoch timestamp,
     * if timestamp is undefined, then epoch timestamp will instead be Infinity.
     * @param epoch
     * @returns {ExpiresAt}
     */
    static fromEpoch(epoch) {
        return new ExpiresAt(epoch);
    }
}
exports.ExpiresAt = ExpiresAt;
