// @swagger-api/apidom-reference/es/dereference/strategies/apidom/index.mjs
import stampit from 'stampit';
import { isElement, cloneDeep, visit } from '@swagger-api/apidom-core';
import DereferenceStrategy from "../DereferenceStrategy.mjs";
import Reference from "../../../Reference.mjs";
import ReferenceSet from "../../../ReferenceSet.mjs";
import ApiDOMDereferenceVisitor from "./visitor.mjs"; // @ts-ignore
const visitAsync = visit[Symbol.for('nodejs.util.promisify.custom')];
const ApiDOMDereferenceStrategy = stampit(DereferenceStrategy, {
  init() {
    this.name = 'apidom';
  },
  methods: {
    canDereference(file) {
      var _file$parseResult;
      return file.mediaType.startsWith('application/vnd.apidom') && isElement((_file$parseResult = file.parseResult) === null || _file$parseResult === void 0 ? void 0 : _file$parseResult.result);
    },
    async dereference(file, options) {
      var _options$dereference$;
      const immutableRefSet = (_options$dereference$ = options.dereference.refSet) !== null && _options$dereference$ !== void 0 ? _options$dereference$ : ReferenceSet();
      const mutableRefsSet = ReferenceSet();
      let refSet = immutableRefSet;
      let reference;

      // determine the initial reference
      if (!immutableRefSet.has(file.uri)) {
        reference = Reference({
          uri: file.uri,
          value: file.parseResult
        });
        immutableRefSet.add(reference);
      } else {
        // pre-computed refSet was provided as configuration option
        reference = immutableRefSet.find(ref => ref.uri === file.uri);
      }

      /**
       * Clone refSet due the dereferencing process being mutable.
       * We don't want to mutate the original refSet and the references.
       */
      if (options.dereference.immutable) {
        immutableRefSet.refs.map(ref => Reference({
          ...ref,
          value: cloneDeep(ref.value)
        })).forEach(ref => mutableRefsSet.add(ref));
        reference = mutableRefsSet.find(ref => ref.uri === file.uri);
        refSet = mutableRefsSet;
      }
      const visitor = ApiDOMDereferenceVisitor({
        reference,
        options
      });
      const dereferencedElement = await visitAsync(refSet.rootRef.value, visitor);

      /**
       * If immutable option is set, replay refs from the refSet.
       */
      if (options.dereference.immutable) {
        mutableRefsSet.refs.filter(ref => ref.uri.startsWith('immutable://')).map(ref => Reference({
          ...ref,
          uri: ref.uri.replace(/^immutable:\/\//, '')
        })).forEach(ref => immutableRefSet.add(ref));
        reference = immutableRefSet.find(ref => ref.uri === file.uri);
        refSet = immutableRefSet;
      }

      /**
       * Release all memory if this refSet was not provided as a configuration option.
       * If provided as configuration option, then provider is responsible for cleanup.
       */
      if (options.dereference.refSet === null) {
        immutableRefSet.clean();
      }
      mutableRefsSet.clean();
      return dereferencedElement;
    }
  }
});
export default ApiDOMDereferenceStrategy;
