// llamaindex@0.2.1/dist/cjs/extractors/prompts.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    defaultKeywordExtractorPromptTemplate: function() {
        return defaultKeywordExtractorPromptTemplate;
    },
    defaultNodeTextTemplate: function() {
        return defaultNodeTextTemplate;
    },
    defaultQuestionAnswerPromptTemplate: function() {
        return defaultQuestionAnswerPromptTemplate;
    },
    defaultSummaryExtractorPromptTemplate: function() {
        return defaultSummaryExtractorPromptTemplate;
    },
    defaultTitleCombinePromptTemplate: function() {
        return defaultTitleCombinePromptTemplate;
    },
    defaultTitleExtractorPromptTemplate: function() {
        return defaultTitleExtractorPromptTemplate;
    }
});
const defaultKeywordExtractorPromptTemplate = ({ contextStr = "", keywords = 5 })=>`${contextStr} 
Give ${keywords} unique keywords for this document. 
Format as comma separated. 
Keywords: `;
const defaultTitleExtractorPromptTemplate = ({ contextStr = "" } = {
    contextStr: ""
})=>`${contextStr}
Give a title that summarizes all of the unique entities, titles or themes found in the context. 
Title: `;
const defaultTitleCombinePromptTemplate = ({ contextStr = "" } = {
    contextStr: ""
})=>`${contextStr} 
Based on the above candidate titles and contents, what is the comprehensive title for this document? 
Title: `;
const defaultQuestionAnswerPromptTemplate = ({ contextStr = "", numQuestions = 5 } = {
    contextStr: "",
    numQuestions: 5
})=>`${contextStr}
Given the contextual informations, generate ${numQuestions} questions this context can provides specific answers to which are unlikely to be found else where. Higher-level summaries of surrounding context may be provideds as well. 
Try using these summaries to generate better questions that this context can answer.
`;
const defaultSummaryExtractorPromptTemplate = ({ contextStr = "" } = {
    contextStr: ""
})=>`${contextStr}
Summarize the key topics and entities of the sections.
Summary: `;
const defaultNodeTextTemplate = ({ metadataStr = "", content = "" } = {})=>`[Excerpt from document]
${metadataStr}
Excerpt:
-----
${content}
-----
`;
