// @gomomento/sdk@1.68.1/dist/src/internal/grpc/headers-interceptor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderInterceptorProvider = exports.Header = void 0;
const grpc_js_1 = require("@grpc/grpc-js");
class Header {
    /**
     * @param {string} name
     * @param {string} value
     */
    constructor(name, value) {
        this.onceOnlyHeaders = ['Agent'];
        this.name = name;
        this.value = value;
    }
}
exports.Header = Header;
class HeaderInterceptorProvider {
    /**
     * @param {Header[]} headers
     */
    constructor(headers) {
        this.headersToAddOnce = headers.filter(header => header.onceOnlyHeaders.includes(header.name));
        this.headersToAddEveryTime = headers.filter(header => !header.onceOnlyHeaders.includes(header.name));
    }
    createHeadersInterceptor() {
        return (options, nextCall) => {
            return new grpc_js_1.InterceptingCall(nextCall(options), {
                start: (metadata, listener, next) => {
                    this.headersToAddEveryTime.forEach(h => metadata.add(h.name, h.value));
                    if (!HeaderInterceptorProvider.areOnlyOnceHeadersSent) {
                        HeaderInterceptorProvider.areOnlyOnceHeadersSent = true;
                        this.headersToAddOnce.forEach(h => metadata.add(h.name, h.value));
                    }
                    next(metadata, {});
                },
            });
        };
    }
}
exports.HeaderInterceptorProvider = HeaderInterceptorProvider;
HeaderInterceptorProvider.areOnlyOnceHeadersSent = false;
