// @swagger-api/apidom-reference/cjs/parse/parsers/asyncapi-json-2/index.cjs
"use strict";

var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
exports.__esModule = true;
exports.default = void 0;
var _stampit = _interopRequireDefault(require("stampit"));
var _ramda = require("ramda");
var _apidomParserAdapterAsyncapiJson = require("@swagger-api/apidom-parser-adapter-asyncapi-json-2");
var _ParserError = _interopRequireDefault(require("../../../errors/ParserError.cjs"));
var _Parser = _interopRequireDefault(require("../Parser.cjs"));
const AsyncApiJson2Parser = (0, _stampit.default)(_Parser.default, {
  props: {
    name: 'asyncapi-json-2',
    fileExtensions: ['.json'],
    mediaTypes: _apidomParserAdapterAsyncapiJson.mediaTypes
  },
  methods: {
    async canParse(file) {
      const hasSupportedFileExtension = this.fileExtensions.length === 0 ? true : this.fileExtensions.includes(file.extension);
      const hasSupportedMediaType = this.mediaTypes.includes(file.mediaType);
      if (!hasSupportedFileExtension) return false;
      if (hasSupportedMediaType) return true;
      if (!hasSupportedMediaType) {
        return (0, _apidomParserAdapterAsyncapiJson.detect)(file.toString());
      }
      return false;
    },
    async parse(file) {
      const source = file.toString();
      try {
        const parserOpts = (0, _ramda.pick)(['sourceMap', 'syntacticAnalysis', 'refractorOpts'], this);
        return await (0, _apidomParserAdapterAsyncapiJson.parse)(source, parserOpts);
      } catch (error) {
        throw new _ParserError.default(`Error parsing "${file.uri}"`, {
          cause: error
        });
      }
    }
  }
});
var _default = exports.default = AsyncApiJson2Parser;
