// @gomomento/sdk@1.68.1/dist/src/webhookutils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateWebhookRequest = exports.RequestValidation = void 0;
const crypto = require("crypto");
var RequestValidation;
(function (RequestValidation) {
    RequestValidation["VALID"] = "valid";
    RequestValidation["INVALID"] = "invalid";
})(RequestValidation = exports.RequestValidation || (exports.RequestValidation = {}));
/**
 * This function is a helper function that can be used to validate that webhook
 * requests are coming from Momento. It is best practice to validate incoming
 * request to a public webhook endpoint
 * @param props {Props}
 */
const validateWebhookRequest = (props) => {
    const hash = crypto.createHmac('SHA3-256', props.signingSecret);
    const hashed = hash.update(props.body).digest('hex');
    console.log('expected', props.signature);
    console.log('actual', hashed);
    if (hashed === props.signature) {
        return RequestValidation.VALID;
    }
    return RequestValidation.INVALID;
};
exports.validateWebhookRequest = validateWebhookRequest;
