// @gomomento/sdk@1.68.1/dist/src/config/logging/default-momento-logger.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultMomentoLogger = exports.DefaultMomentoLoggerFactory = exports.DefaultMomentoLoggerLevel = void 0;
const printf = require("../../internal/vendor/printf/printf");
var DefaultMomentoLoggerLevel;
(function (DefaultMomentoLoggerLevel) {
    DefaultMomentoLoggerLevel[DefaultMomentoLoggerLevel["TRACE"] = 5] = "TRACE";
    DefaultMomentoLoggerLevel[DefaultMomentoLoggerLevel["DEBUG"] = 10] = "DEBUG";
    DefaultMomentoLoggerLevel[DefaultMomentoLoggerLevel["INFO"] = 20] = "INFO";
    DefaultMomentoLoggerLevel[DefaultMomentoLoggerLevel["WARN"] = 30] = "WARN";
    DefaultMomentoLoggerLevel[DefaultMomentoLoggerLevel["ERROR"] = 40] = "ERROR";
})(DefaultMomentoLoggerLevel = exports.DefaultMomentoLoggerLevel || (exports.DefaultMomentoLoggerLevel = {}));
class DefaultMomentoLoggerFactory {
    constructor(level = DefaultMomentoLoggerLevel.INFO) {
        this.level = level;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    getLogger(loggerName) {
        return new DefaultMomentoLogger(loggerName, this.level);
    }
}
exports.DefaultMomentoLoggerFactory = DefaultMomentoLoggerFactory;
class DefaultMomentoLogger {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(loggerName, level) {
        // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
        this.loggerName =
            typeof loggerName === 'string' || loggerName instanceof String
                ? loggerName
                : // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
                    loggerName.constructor.name;
        this.level = level;
    }
    trace(msg, ...args) {
        if (this.level > DefaultMomentoLoggerLevel.TRACE) {
            return;
        }
        this.outputMessage(console.log, DefaultMomentoLoggerLevel.TRACE, msg, args);
    }
    debug(msg, ...args) {
        if (this.level > DefaultMomentoLoggerLevel.DEBUG) {
            return;
        }
        this.outputMessage(console.log, DefaultMomentoLoggerLevel.DEBUG, msg, args);
    }
    info(msg, ...args) {
        if (this.level > DefaultMomentoLoggerLevel.INFO) {
            return;
        }
        this.outputMessage(console.log, DefaultMomentoLoggerLevel.INFO, msg, args);
    }
    warn(msg, ...args) {
        if (this.level > DefaultMomentoLoggerLevel.WARN) {
            return;
        }
        this.outputMessage(console.warn, DefaultMomentoLoggerLevel.WARN, msg, args);
    }
    error(msg, ...args) {
        if (this.level > DefaultMomentoLoggerLevel.ERROR) {
            return;
        }
        this.outputMessage(console.error, DefaultMomentoLoggerLevel.ERROR, msg, args);
    }
    outputMessage(outputFn, level, msg, args) {
        outputFn(printf('[%s] %s (Momento: %s): %s', new Date().toISOString(), DefaultMomentoLoggerLevel[level], this.loggerName, printf(msg, ...args)));
    }
}
exports.DefaultMomentoLogger = DefaultMomentoLogger;
