// @gomomento/generated-types@0.106.1/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.webhook = exports.leaderboard = exports.token = exports.auth = exports.ping = exports.pubsub = exports.control = exports.cache = void 0;
exports.cache = __importStar(require("./cacheclient"));
exports.control = __importStar(require("./controlclient"));
exports.pubsub = __importStar(require("./cachepubsub"));
exports.ping = __importStar(require("./cacheping"));
exports.auth = __importStar(require("./auth"));
exports.token = __importStar(require("./token"));
exports.leaderboard = __importStar(require("./leaderboard"));
exports.webhook = __importStar(require("./webhook"));
