// @gomomento/sdk-core@1.68.1/dist/src/messages/vector-index-info.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.VectorIndexInfo = void 0;
class VectorIndexInfo {
    constructor(name, numDimensions, similarityMetric) {
        this._name = name;
        this._numDimensions = numDimensions;
        this._similarityMetric = similarityMetric;
    }
    get name() {
        return this._name;
    }
    getName() {
        return this._name;
    }
    getNumDimensions() {
        return this._numDimensions;
    }
    get numDimensions() {
        return this._numDimensions;
    }
    getSimilarityMetric() {
        return this._similarityMetric;
    }
    get similarityMetric() {
        return this._similarityMetric;
    }
    toString() {
        return `VectorIndexInfo('${this._name}', ${this._numDimensions}, '${this._similarityMetric}')`;
    }
    equals(other) {
        return (this._name === other._name &&
            this._numDimensions === other._numDimensions &&
            this._similarityMetric === other._similarityMetric);
    }
}
exports.VectorIndexInfo = VectorIndexInfo;
