// llamaindex@0.2.1/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    DEFAULT_CHUNK_OVERLAP: function() {
        return DEFAULT_CHUNK_OVERLAP;
    },
    DEFAULT_CHUNK_OVERLAP_RATIO: function() {
        return DEFAULT_CHUNK_OVERLAP_RATIO;
    },
    DEFAULT_CHUNK_SIZE: function() {
        return DEFAULT_CHUNK_SIZE;
    },
    DEFAULT_CONTEXT_WINDOW: function() {
        return DEFAULT_CONTEXT_WINDOW;
    },
    DEFAULT_NUM_OUTPUTS: function() {
        return DEFAULT_NUM_OUTPUTS;
    },
    DEFAULT_PADDING: function() {
        return DEFAULT_PADDING;
    },
    DEFAULT_SIMILARITY_TOP_K: function() {
        return DEFAULT_SIMILARITY_TOP_K;
    }
});
const DEFAULT_CONTEXT_WINDOW = 3900;
const DEFAULT_NUM_OUTPUTS = 256;
const DEFAULT_CHUNK_SIZE = 1024;
const DEFAULT_CHUNK_OVERLAP = 20;
const DEFAULT_CHUNK_OVERLAP_RATIO = 0.1;
const DEFAULT_SIMILARITY_TOP_K = 2;
const DEFAULT_PADDING = 5;
