// @llamaindex/env@0.0.5/dist/cjs/type.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "InMemoryFileSystem", {
    enumerable: true,
    get: function() {
        return InMemoryFileSystem;
    }
});
const _lodash = /*#__PURE__*/ _interop_require_default(require("lodash"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
class InMemoryFileSystem {
    files = {};
    async writeFile(path, content, options) {
        this.files[path] = _lodash.default.cloneDeep(content);
    }
    async readFile(path) {
        if (!(path in this.files)) {
            throw new Error(`File ${path} does not exist`);
        }
        return _lodash.default.cloneDeep(this.files[path]);
    }
    async access(path) {
        if (!(path in this.files)) {
            throw new Error(`File ${path} does not exist`);
        }
    }
    async mkdir(path) {
        this.files[path] = _lodash.default.get(this.files, path, null);
    }
    async readdir(path) {
        throw new Error("Not implemented");
    }
    async stat(path) {
        throw new Error("Not implemented");
    }
    async readRawFile(path) {
        throw new Error("Not implemented");
    }
}
