// llamaindex@0.2.1/dist/cjs/storage/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    DEFAULT_COLLECTION: function() {
        return DEFAULT_COLLECTION;
    },
    DEFAULT_DOC_STORE_PERSIST_FILENAME: function() {
        return DEFAULT_DOC_STORE_PERSIST_FILENAME;
    },
    DEFAULT_GRAPH_STORE_PERSIST_FILENAME: function() {
        return DEFAULT_GRAPH_STORE_PERSIST_FILENAME;
    },
    DEFAULT_IMAGE_VECTOR_NAMESPACE: function() {
        return DEFAULT_IMAGE_VECTOR_NAMESPACE;
    },
    DEFAULT_INDEX_STORE_PERSIST_FILENAME: function() {
        return DEFAULT_INDEX_STORE_PERSIST_FILENAME;
    },
    DEFAULT_NAMESPACE: function() {
        return DEFAULT_NAMESPACE;
    },
    DEFAULT_PERSIST_DIR: function() {
        return DEFAULT_PERSIST_DIR;
    },
    DEFAULT_VECTOR_STORE_PERSIST_FILENAME: function() {
        return DEFAULT_VECTOR_STORE_PERSIST_FILENAME;
    }
});
const DEFAULT_COLLECTION = "data";
const DEFAULT_PERSIST_DIR = "./storage";
const DEFAULT_INDEX_STORE_PERSIST_FILENAME = "index_store.json";
const DEFAULT_DOC_STORE_PERSIST_FILENAME = "doc_store.json";
const DEFAULT_VECTOR_STORE_PERSIST_FILENAME = "vector_store.json";
const DEFAULT_GRAPH_STORE_PERSIST_FILENAME = "graph_store.json";
const DEFAULT_NAMESPACE = "docstore";
const DEFAULT_IMAGE_VECTOR_NAMESPACE = "images";
