// @gomomento/sdk@1.68.1/dist/src/preview-leaderboard-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreviewLeaderboardClient = void 0;
const leaderboard_data_client_1 = require("./internal/leaderboard-data-client");
const leaderboard_1 = require("./internal/leaderboard");
const index_1 = require("./index");
/**
 * PREVIEW Momento Leaderboard Client
 * WARNING: the API for this client is not yet stable and may change without notice.
 * Please contact Momento if you would like to try this preview.
 *
 * Leaderboard methods return a response object unique to each request.
 * The response object is resolved to a type-safe object of one of several
 * sub-types. See the documentation for each response type for details.
 */
class PreviewLeaderboardClient {
    constructor(props) {
        var _a;
        const configuration = (_a = props.configuration) !== null && _a !== void 0 ? _a : getDefaultLeaderboardConfiguration();
        const propsWithConfig = {
            ...props,
            configuration: configuration,
        };
        this.configuration = configuration;
        this.logger = configuration.getLoggerFactory().getLogger(this);
        this.logger.debug('Creating Momento LeaderboardClient');
        this.dataClient = new leaderboard_data_client_1.LeaderboardDataClient(propsWithConfig, '0'); // only creating one leaderboard client
    }
    close() {
        this.dataClient.close();
        this.configuration.getMiddlewares().forEach(m => {
            if (m.close) {
                m.close();
            }
        });
    }
    /**
     * Creates an instance of LeaderboardClient with floating point scores up until 53 bits of precision.
     */
    leaderboard(cacheName, leaderboardName) {
        return new leaderboard_1.Leaderboard(this.dataClient, cacheName, leaderboardName);
    }
}
exports.PreviewLeaderboardClient = PreviewLeaderboardClient;
function getDefaultLeaderboardConfiguration() {
    return index_1.LeaderboardConfigurations.Laptop.latest();
}
