// @gomomento/sdk@1.68.1/dist/src/preview-vector-index-client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PreviewVectorIndexClient = void 0;
const clients_1 = require("@gomomento/sdk-core/dist/src/internal/clients");
const vector_index_data_client_1 = require("./internal/vector-index-data-client");
const vector_index_control_client_1 = require("./internal/vector-index-control-client");
const index_1 = require("./index");
/**
 * PREVIEW Vector Index Client
 * WARNING: the API for this client is not yet stable and may change without notice.
 *
 * Vector and control methods return a response object unique to each request.
 * The response object is resolved to a type-safe object of one of several
 * sub-types. See the documentation for each response type for details.
 */
class PreviewVectorIndexClient extends clients_1.AbstractVectorIndexClient {
    constructor(props) {
        var _a;
        const configuration = (_a = props.configuration) !== null && _a !== void 0 ? _a : getDefaultVectorIndexConfiguration();
        const propsWithConfiguration = {
            ...props,
            configuration,
        };
        const controlClient = createControlClient(propsWithConfiguration);
        const dataClient = createDataClient(propsWithConfiguration);
        super(controlClient, dataClient);
    }
}
exports.PreviewVectorIndexClient = PreviewVectorIndexClient;
function createControlClient(props) {
    return new vector_index_control_client_1.VectorIndexControlClient({
        configuration: props.configuration,
        credentialProvider: props.credentialProvider,
    });
}
function createDataClient(props) {
    return new vector_index_data_client_1.VectorIndexDataClient(props);
}
function getDefaultVectorIndexConfiguration() {
    return index_1.VectorIndexConfigurations.Laptop.latest();
}
