// portkey-ai@0.1.16/dist/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PORTKEY_PROXY_ENV = exports.PORTKEY_API_KEY_ENV = exports.PORTKEY_BASE_URL = exports.PORTKEY_HEADER_PREFIX = exports.DEFAULT_TIMEOUT = exports.DEFAULT_MAX_RETRIES = exports.INVALID_PORTKEY_MODE = exports.MISSING_MODE_MESSAGE = exports.MISSING_CONFIG_MESSAGE = exports.MISSING_BASE_URL = exports.MISSING_API_KEY_ERROR_MESSAGE = void 0;
exports.MISSING_API_KEY_ERROR_MESSAGE = `No API key found for Portkey.
Please set either the PORTKEY_API_KEY environment variable or \
pass the api_key in the to initialization of Portkey.
API keys can be found or created at Portkey Dashboard \

Here's how you get it:
1. Visit https://app.portkey.ai/
1. Click on your profile icon on the top left
2. From the dropdown menu, click on "Copy API Key"
`;
exports.MISSING_BASE_URL = `No Base url provided. Please provide a valid base url.
For example: https://api.portkey.ai
`;
exports.MISSING_CONFIG_MESSAGE = `The 'config' parameter is not set. Please provide a valid Config object`;
exports.MISSING_MODE_MESSAGE = `The 'mode' parameter is not set. Please provide a valid mode literal.`;
exports.INVALID_PORTKEY_MODE = `Argument of type '{}' cannot be assigned to parameter "mode" of \
type "ModesLiteral | Modes | None"
`;
exports.DEFAULT_MAX_RETRIES = 2;
exports.DEFAULT_TIMEOUT = 60;
exports.PORTKEY_HEADER_PREFIX = "x-portkey-";
exports.PORTKEY_BASE_URL = "https://api.portkey.ai";
exports.PORTKEY_API_KEY_ENV = "PORTKEY_API_KEY";
exports.PORTKEY_PROXY_ENV = "PORTKEY_PROXY";
