// @swagger-api/apidom-reference/es/dereference/strategies/openapi-3-1/index.mjs
import stampit from 'stampit';
import { propEq } from 'ramda';
import { createNamespace, visit, cloneDeep } from '@swagger-api/apidom-core';
import openApi3_1Namespace, { getNodeType, isOpenApi3_1Element, keyMap, mediaTypes } from '@swagger-api/apidom-ns-openapi-3-1';
import DereferenceStrategy from "../DereferenceStrategy.mjs";
import Reference from "../../../Reference.mjs";
import ReferenceSet from "../../../ReferenceSet.mjs";
import OpenApi3_1DereferenceVisitor from "./visitor.mjs"; // @ts-ignore
const visitAsync = visit[Symbol.for('nodejs.util.promisify.custom')];

// eslint-disable-next-line @typescript-eslint/naming-convention
const OpenApi3_1DereferenceStrategy = stampit(DereferenceStrategy, {
  init() {
    this.name = 'openapi-3-1';
  },
  methods: {
    canDereference(file) {
      var _file$parseResult;
      // assert by media type
      if (file.mediaType !== 'text/plain') {
        return mediaTypes.includes(file.mediaType);
      }

      // assert by inspecting ApiDOM
      return isOpenApi3_1Element((_file$parseResult = file.parseResult) === null || _file$parseResult === void 0 ? void 0 : _file$parseResult.result);
    },
    async dereference(file, options) {
      var _options$dereference$;
      const namespace = createNamespace(openApi3_1Namespace);
      const immutableRefSet = (_options$dereference$ = options.dereference.refSet) !== null && _options$dereference$ !== void 0 ? _options$dereference$ : ReferenceSet();
      const mutableRefsSet = ReferenceSet();
      let refSet = immutableRefSet;
      let reference;
      if (!immutableRefSet.has(file.uri)) {
        reference = Reference({
          uri: file.uri,
          value: file.parseResult
        });
        immutableRefSet.add(reference);
      } else {
        // pre-computed refSet was provided as configuration option
        reference = immutableRefSet.find(propEq(file.uri, 'uri'));
      }

      /**
       * Clone refSet due the dereferencing process being mutable.
       * We don't want to mutate the original refSet and the references.
       */
      if (options.dereference.immutable) {
        immutableRefSet.refs.map(ref => Reference({
          ...ref,
          value: cloneDeep(ref.value)
        })).forEach(ref => mutableRefsSet.add(ref));
        reference = mutableRefsSet.find(ref => ref.uri === file.uri);
        refSet = mutableRefsSet;
      }
      const visitor = OpenApi3_1DereferenceVisitor({
        reference,
        namespace,
        options
      });
      const dereferencedElement = await visitAsync(refSet.rootRef.value, visitor, {
        keyMap,
        nodeTypeGetter: getNodeType
      });

      /**
       * If immutable option is set, replay refs from the refSet.
       */
      if (options.dereference.immutable) {
        mutableRefsSet.refs.filter(ref => ref.uri.startsWith('immutable://')).map(ref => Reference({
          ...ref,
          uri: ref.uri.replace(/^immutable:\/\//, '')
        })).forEach(ref => immutableRefSet.add(ref));
        reference = immutableRefSet.find(ref => ref.uri === file.uri);
        refSet = immutableRefSet;
      }

      /**
       * Release all memory if this refSet was not provided as an configuration option.
       * If provided as configuration option, then provider is responsible for cleanup.
       */
      if (options.dereference.refSet === null) {
        immutableRefSet.clean();
      }
      mutableRefsSet.clean();
      return dereferencedElement;
    }
  }
});
export { OpenApi3_1DereferenceVisitor };
export { resolveSchema$refField, resolveSchema$idField, maybeRefractToSchemaElement } from "./util.mjs";
export default OpenApi3_1DereferenceStrategy;
