// llamaindex@0.2.1/dist/cjs/cloud/LlamaCloudRetriever.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "LlamaCloudRetriever", {
    enumerable: true,
    get: function() {
        return LlamaCloudRetriever;
    }
});
const _GlobalsHelper = require("../GlobalsHelper.js");
const _Node = require("../Node.js");
const _ServiceContext = require("../ServiceContext.js");
const _types = require("./types.js");
const _utils = require("./utils.js");
class LlamaCloudRetriever {
    client;
    clientParams;
    retrieveParams;
    projectName = _types.DEFAULT_PROJECT_NAME;
    pipelineName;
    serviceContext;
    resultNodesToNodeWithScore(nodes) {
        return nodes.map((node)=>{
            return {
                // Currently LlamaCloud only supports text nodes
                node: (0, _Node.jsonToNode)(node.node, _Node.ObjectType.TEXT),
                score: node.score
            };
        });
    }
    constructor(params){
        this.clientParams = {
            apiKey: params.apiKey,
            baseUrl: params.baseUrl
        };
        if (params.similarityTopK) {
            params.denseSimilarityTopK = params.similarityTopK;
        }
        this.retrieveParams = params;
        this.pipelineName = params.name;
        if (params.projectName) {
            this.projectName = params.projectName;
        }
        this.serviceContext = params.serviceContext ?? (0, _ServiceContext.serviceContextFromDefaults)();
    }
    async getClient() {
        if (!this.client) {
            this.client = await (0, _utils.getClient)(this.clientParams);
        }
        return this.client;
    }
    async retrieve({ query, parentEvent, preFilters }) {
        const pipelines = await (await this.getClient()).pipeline.searchPipelines({
            projectName: this.projectName,
            pipelineName: this.pipelineName
        });
        if (pipelines.length !== 1 && !pipelines[0].id) {
            throw new Error(`No pipeline found with name ${this.pipelineName} in project ${this.projectName}`);
        }
        const results = await (await this.getClient()).pipeline.runSearch(pipelines[0].id, {
            ...this.retrieveParams,
            query,
            searchFilters: preFilters
        });
        const nodes = this.resultNodesToNodeWithScore(results.retrievalNodes);
        if (this.serviceContext.callbackManager.onRetrieve) {
            this.serviceContext.callbackManager.onRetrieve({
                query,
                nodes,
                event: _GlobalsHelper.globalsHelper.createEvent({
                    parentEvent,
                    type: "retrieve"
                })
            });
        }
        return nodes;
    }
    getServiceContext() {
        return this.serviceContext;
    }
}
